### PR TITLE
Scalafix-core compiled in Scala 3

### PR DESCRIPTION
**Pull request by rvacaru**
This PR is the first step towards adapting ExplicitResultTypes for scala3 #1583 and part of the plan for scala3 support #1316. The idea is to cross compile all scalafix modules to scala3 and then introduce a new `ExplicitResultTypes` implementation for scala3.

Screenshot 2022-07-22 at 12 03 53

In this first step we're cross compiling `scalafix-core` to scala3

### CODE REVIEW

File: scalafix-core/src/main/scala-3/scalafix/util/Compat.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package scalafix.util
+
+import scala.collection.immutable.IndexedSeq
+
+object Compat {
+  type View[T] = collection.SeqView[T] //, IndexedSeq[T]]
+  type SeqView[T] = collection.SeqView[T] //, IndexedSeq[T]]
```
Comment (rvacaru) [inline scalafix-core/src/main/scala-3/scalafix/util/Compat.scala:7]: if I provide `[T], IndexedSeq[T]]` it fails compiling saying there are too many type args and it only expects one [A] type arg.. would you know how to fix this?

Comment (rvacaru) [reply scalafix-core/src/main/scala-3/scalafix/util/Compat.scala:7]: Thanks @armanbilge, now it's more clear

File: build.sbt [hunk 2]
```diff
@@ -64,18 +64,34 @@ lazy val core = projectMatrix
...
     moduleName := "scalafix-core",
     buildInfoSettingsForCore,
-    libraryDependencies ++= List(
-      scalameta,
-      googleDiff,
-      collectionCompat
+    scalacOptions --= (
+      if (isScala3.value)
+        Seq("-P:semanticdb:synthetics:on")
+      else Nil
```
Comment (bjaglin) [inline build.sbt:70]: this is repeated in many places, can we move it to a settings block covering all current usages?

File: build.sbt [hunk 3]
```diff
@@ -64,18 +64,34 @@ lazy val core = projectMatrix
...
     ),
+    libraryDependencies ++= {
+      CrossVersion.partialVersion(scalaVersion.value) match {
+        case Some((3, _)) => List(
+          scalameta
+            .exclude("com.lihaoyi", "sourcecode_2.13")
+            .exclude("org.scala-lang.modules", "scala-collection-compat_2.13"),
+          googleDiff
```
Comment (bjaglin) [inline build.sbt:78]: extract this common one to another block

File: build.sbt [hunk 4]
```diff
@@ -309,6 +325,11 @@ lazy val unit = projectMatrix
     }
   )
   .defaultAxes(VirtualAxis.jvm)
+  .jvmPlatform(
+    scalaVersions = Seq(scala3),
+    axisValues = Seq(TargetAxis(scala3)),
+    settings = Seq()
+  )
```
Comment (bjaglin) [inline build.sbt:332]: let's remove this for now, focusing on getting a green build

File: project/Dependencies.scala [hunk 5]
```diff
@@ -53,7 +53,7 @@ object Dependencies {
   val nailgunServer = "com.martiansoftware" % "nailgun-server" % nailgunV
   val scalaXml = "org.scala-lang.modules" %% "scala-xml" % scalaXmlV
   val scalaXml211 = "org.scala-lang.modules" %% "scala-xml" % scalaXml211V
-  val scalameta = "org.scalameta" %% "scalameta" % scalametaV
+  val scalameta = ("org.scalameta" %% "scalameta" % scalametaV).cross(CrossVersion.for3Use2_13)
```
Comment (bjaglin) [inline project/Dependencies.scala:56]:
```suggestion
  // https://github.com/scalameta/scalameta/issues/2485
  val scalameta = ("org.scalameta" %% "scalameta" % scalametaV).cross(CrossVersion.for3Use2_13)
```

File: scalafix-core/src/main/scala-2/scalafix/internal/util/MetaconfigCompatMacros.scala [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
 package scalafix.internal.util
```
Comment (bjaglin) [inline scalafix-core/src/main/scala-2/scalafix/internal/util/MetaconfigCompatMacros.scala:2]: It looks like https://github.com/scalacenter/scalafix/pull/1562 was only partially reverted - why? Thi seems to be used in the scala-2 version of `ConfigRulePatches` but should not longer be the case (so you won't need a separate scala-3 version).

For visibility, let's spin off the revert of https://github.com/scalacenter/scalafix/pull/1562 in a separate PR with the rationale (major clients have been released and scala-steward is back so the risk of using an old scalafix-cli with a rule compiled against a recent scalafix-core is limited).

File: project/ScalafixBuild.scala [hunk 7]
```diff
@@ -184,7 +184,10 @@ object ScalafixBuild extends AutoPlugin with GhpagesKeys {
     Compile / unmanagedSourceDirectories ++= {
       val sourceDir = (Compile / sourceDirectory).value
       CrossVersion.partialVersion(scalaVersion.value) match {
-        case Some((2, n)) if n >= 12 => Seq(sourceDir / "scala-2.12+")
+        case Some((2, n)) if n == 13 => Seq(sourceDir / "scala-2.13")
+        case Some((2, n)) if n == 12 => Seq(sourceDir / "scala-2.12")
+        case Some((2, n)) if n == 11 => Seq(sourceDir / "scala-2.11")
+        case Some((3, _))  => Seq(sourceDir / "scala-3")
```
Comment (bjaglin) [inline project/ScalafixBuild.scala:190]: AFAIK, all these are built-in in sbt, why did you add them here?

File: project/ScalafixBuild.scala [hunk 8]
```diff
@@ -184,7 +184,10 @@ object ScalafixBuild extends AutoPlugin with GhpagesKeys {
     Compile / unmanagedSourceDirectories ++= {
       val sourceDir = (Compile / sourceDirectory).value
       CrossVersion.partialVersion(scalaVersion.value) match {
-        case Some((2, n)) if n >= 12 => Seq(sourceDir / "scala-2.12+")
```
Comment (bjaglin) [inline project/ScalafixBuild.scala:187]: I still see files under these directories, so the removal of this makes them dead code (and probably compilation errors). This avoid duplication between 2.12/2.13 source directories, so why not keep it?

File: scalafix-core/src/main/scala-3/scalafix/internal/util/PrettyType.scala [hunk 9]
```diff
@@ -0,0 +1,732 @@
+package scalafix.internal.util
```
Comment (bjaglin) [inline scalafix-core/src/main/scala-3/scalafix/internal/util/PrettyType.scala:1]: These new files in `scala-3` are HUGE - so much duplicated code will be very costly to keep in sync. Instead try to keep the common code in `src/mains/scala` and extract the differences in a smaller file (`Compat.scala`).

File: build.sbt [hunk 10]
```diff
@@ -64,18 +64,34 @@ lazy val core = projectMatrix
...
-      googleDiff,
-      collectionCompat
+    scalacOptions --= (
+      if (isScala3.value)
+        Seq("-P:semanticdb:synthetics:on")
+      else Nil
     ),
+    libraryDependencies ++= {
+      CrossVersion.partialVersion(scalaVersion.value) match {
+        case Some((3, _)) => List(
```
Comment (bjaglin) [inline build.sbt:74]: `if (isScala3.value)` is more readable if you just to branch based on the major version

File: scalafix-core/src/main/scala/scalafix/internal/util/Pretty.scala [hunk 11]
```diff
@@ -5,6 +5,7 @@ import scala.meta.Lit
 import org.typelevel.paiges.Doc
 import scalafix.internal.v1.Types
 import scalafix.v1._
+import meta.XtensionSyntax
```
Comment (bjaglin) [inline scalafix-core/src/main/scala/scalafix/internal/util/Pretty.scala:8]: is meta the root package or is that a relative import?

File: scalafix-core/src/main/scala/scalafix/util/TokenClasses.scala [hunk 12]
```diff
@@ -3,6 +3,7 @@ package scalafix.util
 import scala.meta.classifiers.Classifier
 import scala.meta.tokens.Token
 import scala.meta.tokens.Token._
+import meta.classifiers.XtensionClassifiable
```
Comment (bjaglin) [inline scalafix-core/src/main/scala/scalafix/util/TokenClasses.scala:6]: is meta the root package or is that a relative import?

**Comment by bjaglin** [comment]
Another note: since `core3` is an orphan modules not used by any other, I don't think it's covered by CI, so you should add an explicit `unit3/compile` in https://github.com/scalacenter/scalafix/blob/59ed8c053a79cadfe3ef2f83a10561df7170a192/project/ScalafixBuild.scala#L131-L134
This will apply for all xxx3 modules until we move unit2_12Target3 to unit3.

File: scalafix-core/src/main/scala/scalafix/util/TokenClasses.scala (hunk 12 above)
Comment (rvacaru) [reply scalafix-core/src/main/scala/scalafix/util/TokenClasses.scala:6]: it's a relative import, will fix all of them

File: project/Dependencies.scala (hunk 5 above)
Comment (rvacaru) [reply project/Dependencies.scala:56]: added

File: scalafix-core/src/main/scala-2/scalafix/internal/util/MetaconfigCompatMacros.scala (hunk 6 above)
Comment (rvacaru) [reply scalafix-core/src/main/scala-2/scalafix/internal/util/MetaconfigCompatMacros.scala:2]: Here's the separate PR for the revert, feel free to edit the description if needed: https://github.com/scalacenter/scalafix/pull/1632

Comment (bjaglin) [reply scalafix-core/src/main/scala-2/scalafix/internal/util/MetaconfigCompatMacros.scala:2]: merged, you can rebase

File: build.sbt (hunk 2 above)
Comment (rvacaru) [reply build.sbt]: factored out

File: scalafix-core/src/main/scala-3/scalafix/internal/util/PrettyType.scala (hunk 9 above)
Comment (rvacaru) [reply scalafix-core/src/main/scala-3/scalafix/internal/util/PrettyType.scala]: refactored similarly as Compat, defining different types for different scala versions

File: scalafix-core/src/main/scala-3/scalafix/util/TokenList.scala [hunk 13]
```diff
@@ -0,0 +1,77 @@
+package scalafix.util
+
+import scala.meta.tokens.Token
+import scala.meta.tokens.Tokens
+import scala.meta.classifiers.XtensionClassifiable
+
+import scalafix.util.Compat._
```
Comment (bjaglin) [inline scalafix-core/src/main/scala-3/scalafix/util/TokenList.scala:7]: the idea of Compat classes is that
- they have different implementations depending on the version
- they are referenced in code common to all versions

In that case, can't you try to keep `TokenList` generic? I see the original one was moved to `scala-2` without changes - instead it should remain in `scala` with minor modifications.

File: scalafix-core/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 14]
```diff
@@ -172,7 +173,7 @@ class PrettyType private (
         Nil,
         Pat.Var(Term.Name(info.displayName)) :: Nil,
         None,
-        q"???"
+        "???".parse[Term].get
```
Comment (bjaglin) [inline scalafix-core/src/main/scala/scalafix/internal/util/PrettyType.scala:176]: why did you have to change that? The `.get` is suspicious, what about
```suggestion
        Term.Name("???")
```

File: project/ScalafixBuild.scala [hunk 15]
```diff
@@ -130,6 +130,7 @@ object ScalafixBuild extends AutoPlugin with GhpagesKeys {
     },
     commands += Command.command("ci-3") { s =>
       "unit2_12Target3/test" ::
+      "unit3/compile" ::
```
Comment (bjaglin) [inline project/ScalafixBuild.scala:133]:
```suggestion
      "core3/compile" ::
```

File: scalafix-core/src/main/scala-2/scalafix/internal/config/ConfigRulePatches.scala [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    addGlobalImports: List[AddGlobalImport] = Nil,
+    removeGlobalImports: List[RemoveGlobalImport] = Nil
+) {
+  def all: List[Patch] =
+    replaceSymbols ++ addGlobalImports ++ removeGlobalImports
+}
+
+object ConfigRulePatches {
+  implicit val surface: Surface[ConfigRulePatches] =
+    MetaconfigCompatMacros.deriveSurfaceOrig[ConfigRulePatches]
```
Comment (bjaglin) [inline scalafix-core/src/main/scala-2/scalafix/internal/config/ConfigRulePatches.scala:21]: something went wrong with your rebase - this file and `MetaconfigCompatMacros` shouldn't be there, and `ConfigRulePatches` should remain in `scala`

File: build.sbt [hunk 17]
```diff
@@ -26,6 +26,12 @@ def inferJavaHome() = {
   Some(actualHome)
 }
 
+lazy val scala3ScalacOption = Def.setting (
```
Comment (bjaglin) [inline build.sbt:29]: there is a much cleaner way to do this:
- since the decision must be taken at the project level and not at the build level, remove https://github.com/scalacenter/scalafix/blob/59ed8c053a79cadfe3ef2f83a10561df7170a192/build.sbt#L8
- add the opposite of this (i.e. adding `-P:semanticdb:synthetics:on` if not scala3) in https://github.com/scalacenter/scalafix/blob/59ed8c053a79cadfe3ef2f83a10561df7170a192/project/ScalafixBuild.scala#L180
- you can now get rid of all the customizations!

File: build.sbt [hunk 18]
```diff
@@ -64,18 +70,31 @@ lazy val core = projectMatrix
...
     buildInfoSettingsForCore,
-    libraryDependencies ++= List(
-      scalameta,
-      googleDiff,
-      collectionCompat
-    ),
+    scalacOptions --= (scala3ScalacOption.value),
+    libraryDependencies += {
+      googleDiff
+    },
```
Comment (bjaglin) [inline build.sbt:76]:
```suggestion
    libraryDependencies += googleDiff
```

**Review by bjaglin — COMMENTED** [review]
Last round of comments, and we'll be good to go (provided `core3/compile` is green!)

File: build.sbt (hunk 17 above)
Comment (rvacaru) [reply build.sbt]: having the reverse logic of adding the option only when needed is more readable, i like it!

File: scalafix-core/src/main/scala-3/scalafix/util/TokenList.scala (hunk 13 above)
Comment (rvacaru) [reply scalafix-core/src/main/scala-3/scalafix/util/TokenList.scala]: my bad, looking at it again the problem was that explicitly defining types `View[T]` in the below methods prevented the `Compat` classes to do their work:

```
  def leadingSpaces(token: Token): View[T] =
    leading(token).takeWhile(_.is[Token.Space])

  def trailingSpaces(token: Token): View[T] =
    trailing(token).takeWhile(_.is[Token.Space])

```

by removing the types it now compiles for the different scala versions

File: scalafix-core/src/main/scala-3/scalafix/util/Compat.scala [hunk 19]
```diff
@@ -0,0 +1,9 @@
+package scalafix.util
+
+import scala.collection.immutable.IndexedSeq
+
+object Compat {
```
Comment (bjaglin) [inline scalafix-core/src/main/scala-3/scalafix/util/Compat.scala:5]: unreferenced?

Comment (rvacaru) [reply scalafix-core/src/main/scala-3/scalafix/util/Compat.scala:3]: I don't think so, also for the other `Compat` classes (at least in VS code) I don't get any references on the class level. Moreover if I remove the class I get a compilation error:
```
sbt:scalafix> core3/compile
[info] compiling 125 Scala sources to /Users/razvan.vacaru@TMNL.nl/srepo/scalafix/scalafix-core/target/jvm-3/classes ...
[error] -- [E008] Not Found Error: /Users/razvan.vacaru@TMNL.nl/srepo/scalafix/scalafix-core/src/main/scala/scalafix/util/TokenList.scala:7:21 
[error] 7 |import scalafix.util.Compat._
[error]   |       ^^^^^^^^^^^^^^^^^^^^
[error]   |value Compat is not a member of scalafix.util, but could be made available as an extension method.
[error]   |
[error]   |The following import might make progress towards fixing the problem:
[error]   |
[error]   |  import sourcecode.Text.generate
[error]   |
[error] one error found
[error] (core3 / Compile / compileIncremental) Compilation failed
```

Comment (bjaglin) [reply scalafix-core/src/main/scala-3/scalafix/util/Compat.scala:3]: oh, sorry, I thought you had introduced a new Compat class, but this is just the scala-3 variant of an existing one :+1:

**Review by bjaglin — APPROVED** [review]
